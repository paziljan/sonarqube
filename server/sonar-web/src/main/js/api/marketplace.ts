/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { checkStatus, corsRequest, getJSON, parseJSON, postJSON } from '../helpers/request';
import throwGlobalError from '../app/utils/throwGlobalError';

export interface Edition {
  key: string;
  name: string;
  desc: string;
  more_link: string;
  request_license_link: string;
  download_link: string;
}

export interface EditionsPerVersion {
  [version: string]: Edition[];
}

export interface EditionStatus {
  currentEditionKey?: string;
  nextEditionKey?: string;
  installationStatus:
    | 'NONE'
    | 'AUTOMATIC_IN_PROGRESS'
    | 'MANUAL_IN_PROGRESS'
    | 'AUTOMATIC_READY'
    | 'AUTOMATIC_FAILURE'
    | 'UNINSTALL_IN_PROGRESS';
}

export function getEditionStatus(): Promise<EditionStatus> {
  return getJSON('/api/editions/status').catch(throwGlobalError);
}

export function getEditionsList(url: string): Promise<EditionsPerVersion> {
  return corsRequest(url)
    .submit()
    .then(checkStatus)
    .then(parseJSON);
}

export function getLicensePreview(data: {
  license: string;
}): Promise<{
  nextEditionKey: string;
  previewStatus: 'NO_INSTALL' | 'AUTOMATIC_INSTALL' | 'MANUAL_INSTALL';
}> {
  return postJSON('/api/editions/preview', data).catch(throwGlobalError);
}

export function applyLicense(data: { license: string }): Promise<EditionStatus> {
  return postJSON('/api/editions/apply_license', data).catch(throwGlobalError);
}

export function uninstallEdition(): Promise<void | Response> {
  return postJSON('/api/editions/uninstall').catch(throwGlobalError);
}
