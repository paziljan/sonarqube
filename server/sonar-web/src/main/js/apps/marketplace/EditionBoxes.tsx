/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import EditionBox from './components/EditionBox';
import LicenseEditionForm from './components/LicenseEditionForm';
import UninstallEditionForm from './components/UninstallEditionForm';
import { Edition, EditionStatus, getEditionsList } from '../../api/marketplace';
import { getEditionsForVersion } from './utils';
import { translate } from '../../helpers/l10n';

export interface Props {
  editionStatus?: EditionStatus;
  editionsUrl: string;
  sonarqubeVersion: string;
  updateCenterActive: boolean;
  updateEditionStatus: (editionStatus: EditionStatus) => void;
}

interface State {
  editions?: Edition[];
  editionsError: boolean;
  loading: boolean;
  installEdition?: Edition;
  openUninstallForm: boolean;
}

export default class EditionBoxes extends React.PureComponent<Props, State> {
  mounted: boolean;
  state: State = { editionsError: false, loading: true, openUninstallForm: false };

  componentDidMount() {
    this.mounted = true;
    this.fetchEditions();
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  fetchEditions = () => {
    this.setState({ loading: true });
    getEditionsList(this.props.editionsUrl).then(
      editionsPerVersion => {
        if (this.mounted) {
          this.setState({
            loading: false,
            editions: getEditionsForVersion(editionsPerVersion, this.props.sonarqubeVersion),
            editionsError: false
          });
        }
      },
      () => {
        if (this.mounted) {
          this.setState({ editionsError: true, loading: false });
        }
      }
    );
  };

  handleOpenLicenseForm = (edition: Edition) => this.setState({ installEdition: edition });
  handleCloseLicenseForm = () => this.setState({ installEdition: undefined });

  handleOpenUninstallForm = () => this.setState({ openUninstallForm: true });
  handleCloseUninstallForm = () => this.setState({ openUninstallForm: false });

  render() {
    const { editionStatus } = this.props;
    const { editions, editionsError, loading, installEdition, openUninstallForm } = this.state;
    if (loading) {
      return <i className="big-spacer-bottom spinner" />;
    }

    if (!editions || editionsError) {
      return (
        <div className="spacer-bottom marketplace-editions">
          <span className="alert alert-info">
            <FormattedMessage
              defaultMessage={translate('marketplace.editions_unavailable')}
              id="marketplace.editions_unavailable"
              values={{
                url: (
                  <a href="https://www.sonarsource.com" target="_blank">
                    SonarSource.com
                  </a>
                )
              }}
            />
          </span>
        </div>
      );
    }

    return (
      <div className="spacer-bottom marketplace-editions">
        {editions.map(edition => (
          <EditionBox
            edition={edition}
            editionStatus={editionStatus}
            key={edition.key}
            onInstall={this.handleOpenLicenseForm}
            onUninstall={this.handleOpenUninstallForm}
          />
        ))}

        {installEdition && (
          <LicenseEditionForm
            edition={installEdition}
            editions={editions}
            onClose={this.handleCloseLicenseForm}
            updateEditionStatus={this.props.updateEditionStatus}
          />
        )}

        {openUninstallForm &&
        editionStatus &&
        editionStatus.currentEditionKey && (
          <UninstallEditionForm
            edition={editions.find(edition => edition.key === editionStatus.currentEditionKey)}
            editionStatus={editionStatus}
            onClose={this.handleCloseUninstallForm}
            updateEditionStatus={this.props.updateEditionStatus}
          />
        )}
      </div>
    );
  }
}
