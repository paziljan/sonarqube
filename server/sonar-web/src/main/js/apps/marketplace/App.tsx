/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import * as PropTypes from 'prop-types';
import { sortBy, uniqBy } from 'lodash';
import Helmet from 'react-helmet';
import Header from './Header';
import EditionsStatusNotif from './components/EditionsStatusNotif';
import EditionBoxes from './EditionBoxes';
import Footer from './Footer';
import PendingActions from './PendingActions';
import PluginsList from './PluginsList';
import Search from './Search';
import {
  getAvailablePlugins,
  getInstalledPluginsWithUpdates,
  getPendingPlugins,
  getPluginUpdates,
  Plugin,
  PluginPending
} from '../../api/plugins';
import { EditionStatus, getEditionStatus } from '../../api/marketplace';
import { RawQuery } from '../../helpers/query';
import { translate } from '../../helpers/l10n';
import { filterPlugins, parseQuery, Query, serializeQuery } from './utils';

export interface Props {
  editionsUrl: string;
  location: { pathname: string; query: RawQuery };
  sonarqubeVersion: string;
  updateCenterActive: boolean;
}

interface State {
  editionStatus?: EditionStatus;
  loading: boolean;
  pending: {
    installing: PluginPending[];
    updating: PluginPending[];
    removing: PluginPending[];
  };
  plugins: Plugin[];
}

export default class App extends React.PureComponent<Props, State> {
  mounted: boolean;

  static contextTypes = {
    router: PropTypes.object.isRequired
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      loading: true,
      pending: {
        installing: [],
        updating: [],
        removing: []
      },
      plugins: []
    };
  }

  componentDidMount() {
    this.mounted = true;
    this.fetchPendingPlugins();
    this.fetchEditionStatus();
    this.fetchQueryPlugins();
  }

  componentDidUpdate(prevProps: Props) {
    if (prevProps.location.query.filter !== this.props.location.query.filter) {
      this.fetchQueryPlugins();
    }
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  fetchQueryPlugins = () => {
    const query = parseQuery(this.props.location.query);
    if (query.filter === 'updates') {
      this.fetchUpdatesOnly();
    } else {
      this.fetchAllPlugins();
    }
  };

  fetchAllPlugins = () => {
    this.setState({ loading: true });
    Promise.all([getInstalledPluginsWithUpdates(), getAvailablePlugins()]).then(
      ([installed, available]) => {
        if (this.mounted) {
          this.setState({
            loading: false,
            plugins: sortBy(uniqBy([...installed, ...available.plugins], 'key'), 'name')
          });
        }
      },
      () => {
        if (this.mounted) {
          this.setState({ loading: false });
        }
      }
    );
  };

  fetchUpdatesOnly = () => {
    this.setState({ loading: true });
    getPluginUpdates().then(
      plugins => {
        if (this.mounted) {
          this.setState({ loading: false, plugins });
        }
      },
      () => {
        if (this.mounted) {
          this.setState({ loading: false });
        }
      }
    );
  };

  fetchPendingPlugins = () =>
    getPendingPlugins().then(
      pending => {
        if (this.mounted) {
          this.setState({ pending });
        }
      },
      () => {}
    );

  fetchEditionStatus = () =>
    getEditionStatus().then(
      editionStatus => {
        if (this.mounted) {
          this.updateEditionStatus(editionStatus);
          setTimeout(this.fetchEditionStatus, 5000);
        }
      },
      () => {}
    );

  updateEditionStatus = (editionStatus: EditionStatus) =>
    this.setState({ editionStatus: editionStatus });

  updateQuery = (newQuery: Partial<Query>) => {
    const query = serializeQuery({ ...parseQuery(this.props.location.query), ...newQuery });
    this.context.router.push({ pathname: this.props.location.pathname, query });
  };

  render() {
    const { editionStatus, loading, plugins, pending } = this.state;
    const query = parseQuery(this.props.location.query);
    const filteredPlugins = query.search ? filterPlugins(plugins, query.search) : plugins;
    return (
      <div className="page page-limited" id="marketplace-page">
        <Helmet title={translate('marketplace.page')} />
        <div className="marketplace-notifs">
          {editionStatus && (
            <EditionsStatusNotif
              editionStatus={editionStatus}
              updateEditionStatus={this.updateEditionStatus}
            />
          )}
          <PendingActions refreshPending={this.fetchPendingPlugins} pending={pending} />
        </div>
        <Header />
        <EditionBoxes
          editionStatus={editionStatus}
          editionsUrl={this.props.editionsUrl}
          sonarqubeVersion={this.props.sonarqubeVersion}
          updateCenterActive={this.props.updateCenterActive}
          updateEditionStatus={this.updateEditionStatus}
        />
        <Search
          query={query}
          updateCenterActive={this.props.updateCenterActive}
          updateQuery={this.updateQuery}
        />
        {loading && <i className="spinner" />}
        {!loading && (
          <PluginsList
            plugins={filteredPlugins}
            pending={pending}
            refreshPending={this.fetchPendingPlugins}
            updateQuery={this.updateQuery}
          />
        )}
        {!loading && <Footer total={filteredPlugins.length} />}
      </div>
    );
  }
}
