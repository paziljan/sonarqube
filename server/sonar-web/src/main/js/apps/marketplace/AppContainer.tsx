/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { connect } from 'react-redux';
import App from './App';
import { getAppState, getGlobalSettingValue } from '../../store/rootReducer';
import './style.css';

const mapStateToProps = (state: any) => ({
  editionsUrl: (getGlobalSettingValue(state, 'sonar.editions.jsonUrl') || {}).value,
  sonarqubeVersion: getAppState(state).version,
  updateCenterActive:
    (getGlobalSettingValue(state, 'sonar.updatecenter.activate') || {}).value === 'true'
});

export default connect(mapStateToProps)(App as any);
