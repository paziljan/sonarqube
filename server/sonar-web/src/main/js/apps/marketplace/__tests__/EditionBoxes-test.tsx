/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import EditionBoxes from '../EditionBoxes';
import { EditionStatus } from '../../../api/marketplace';

const DEFAULT_STATUS: EditionStatus = {
  currentEditionKey: 'foo',
  nextEditionKey: '',
  installationStatus: 'NONE'
};

const DEFAULT_EDITIONS = [
  {
    key: 'foo',
    name: 'Foo',
    desc: 'Foo desc',
    download_link: 'download_url',
    more_link: 'more_url',
    request_license_link: 'license_url'
  },
  {
    key: 'bar',
    name: 'Bar',
    desc: 'Bar desc',
    download_link: 'download_url',
    more_link: 'more_url',
    request_license_link: 'license_url'
  }
];

it('should display the edition boxes', () => {
  const wrapper = getWrapper({ editions: DEFAULT_EDITIONS, loading: true });
  expect(wrapper).toMatchSnapshot();
  wrapper.setProps({ loading: false });
  expect(wrapper).toMatchSnapshot();
});

it('should display an error message', () => {
  const wrapper = getWrapper();
  expect(wrapper).toMatchSnapshot();
});

it('should open the license form', () => {
  const wrapper = getWrapper({ editions: DEFAULT_EDITIONS });
  (wrapper.instance() as EditionBoxes).handleOpenLicenseForm(DEFAULT_EDITIONS[0]);
  expect(wrapper.find('LicenseEditionForm').exists()).toBeTruthy();
});

function getWrapper(props = {}) {
  return shallow(
    <EditionBoxes
      loading={false}
      editionStatus={DEFAULT_STATUS}
      editionsUrl=""
      sonarqubeVersion="6.7.5"
      updateCenterActive={true}
      updateEditionStatus={jest.fn()}
      {...props}
    />
  );
}
