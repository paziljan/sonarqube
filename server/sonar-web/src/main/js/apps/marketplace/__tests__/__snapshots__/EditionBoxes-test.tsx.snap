// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display an error message 1`] = `
<div
  className="spacer-bottom marketplace-editions"
>
  <span
    className="alert alert-info"
  >
    <FormattedMessage
      defaultMessage="marketplace.editions_unavailable"
      id="marketplace.editions_unavailable"
      values={
        Object {
          "url": <a
            href="https://www.sonarsource.com"
            target="_blank"
        >
            SonarSource.com
        </a>,
        }
      }
    />
  </span>
</div>
`;

exports[`should display the edition boxes 1`] = `
<i
  className="big-spacer-bottom spinner"
/>
`;

exports[`should display the edition boxes 2`] = `
<div
  className="spacer-bottom marketplace-editions"
>
  <EditionBox
    edition={
      Object {
        "desc": "Foo desc",
        "download_link": "download_url",
        "key": "foo",
        "more_link": "more_url",
        "name": "Foo",
        "request_license_link": "license_url",
      }
    }
    editionStatus={
      Object {
        "currentEditionKey": "foo",
        "installationStatus": "NONE",
        "nextEditionKey": "",
      }
    }
    onInstall={[Function]}
    onUninstall={[Function]}
  />
  <EditionBox
    edition={
      Object {
        "desc": "Bar desc",
        "download_link": "download_url",
        "key": "bar",
        "more_link": "more_url",
        "name": "Bar",
        "request_license_link": "license_url",
      }
    }
    editionStatus={
      Object {
        "currentEditionKey": "foo",
        "installationStatus": "NONE",
        "nextEditionKey": "",
      }
    }
    onInstall={[Function]}
    onUninstall={[Function]}
  />
</div>
`;
